### PR TITLE
Exaqute/base moment

### PR DESCRIPTION
**Pull request by akodakkal**
Compute base moment process added to compute the moment of forces about a reference point and write them out.

### CODE REVIEW

File: kratos/geometries/geometry.h [hunk 1]
```diff
@@ -1386,7 +1386,10 @@ class Geometry
         array_1d<double, 3> normal = Normal(rPointLocalCoordinates);
         const double norm_normal = norm_2(normal);
         if (norm_normal > std::numeric_limits<double>::epsilon()) normal /= norm_normal;
-        else KRATOS_ERROR << "ERROR: The normal norm is zero or almost zero. Norm. normal: " << norm_normal << std::endl;
+        else {
+            std::cout << rPointLocalCoordinates << std::endl;
```
Comment (riccardotosi) [inline kratos/geometries/geometry.h]: I think you pushed this change by error. You can `git checkout` the `master` version of the file.

File: applications/MeshingApplication/custom_processes/mmg_process.cpp [hunk 2]
```diff
@@ -1227,7 +1228,7 @@ void MmgProcess<TMMGLibrary>::CleanSuperfluousNodes()
     const auto it_elem_begin = r_elements_array.begin();
 
     // Saving the nodes that belong to an element
-    #pragma omp parallel for
+    // #pragma omp parallel for
```
Comment (riccardotosi) [inline applications/MeshingApplication/custom_processes/mmg_process.cpp]: I think you pushed this change by error. You can `git checkout` the `master` version of the file.

File: applications/MeshingApplication/custom_processes/mmg_process.cpp [hunk 3]
```diff
@@ -1239,6 +1240,30 @@ void MmgProcess<TMMGLibrary>::CleanSuperfluousNodes()
...
+    }
+
+    for (auto  id_cond : list_of_ids_to_remove) {
+        // KRATOS_INFO("MmgProcess") << "Removing condition #" << id_cond << std::endl;
+        mrThisModelPart.RemoveConditionFromAllLevels(id_cond);
+    }
+
+
+    // mrThisModelPart.RemoveConditionFromAllLevels(TO_ERASE);
+
```
Comment (riccardotosi) [inline applications/MeshingApplication/custom_processes/mmg_process.cpp]: Why is this needed in a base moment process?

File: applications/MeshingApplication/custom_processes/mmg_process.cpp [hunk 4]
```diff
@@ -1211,6 +1211,7 @@ void MmgProcess<TMMGLibrary>::MarkConditionsSubmodelParts(ModelPart& rModelPart)
 /***********************************************************************************/
 
 template<MMGLibrary TMMGLibrary>
+
```
Comment (riccardotosi) [inline applications/MeshingApplication/custom_processes/mmg_process.cpp]:
```suggestion
```
But it is better to `git checkout` the `master` version of the file.

File: applications/FluidDynamicsApplication/python_scripts/compute_embedded_base_moment_process.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+def Factory(settings, model):
+    if(type(settings) != KratosMultiphysics.Parameters):
+        raise Exception("expected input shall be a Parameters object, encapsulating a json string")
+
+    return ComputeEmbeddedBaseMomentProcess(model, settings["Parameters"])
+
+class ComputeEmbeddedBaseMomentProcess(ComputeBaseMomentProcess):
+    """
+    The specific implementation for the output of body fitted drag forces
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/python_scripts/compute_embedded_base_moment_process.py]:
```suggestion
    The specific implementation for the output of embedded fitted drag forces
```

File: applications/FluidDynamicsApplication/python_scripts/compute_body_fitted_base_moment_process.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+        header += '# Time Mx My Mz\n'
+        return header
+
+    def _PrintToScreen(self, result_msg):
+        KratosMultiphysics.Logger.PrintInfo("ComputeBodyFittedBaseMomentProcess","BODY FITTED MOMENT RESULTS:")
+        KratosMultiphysics.Logger.PrintInfo("ComputeBodyFittedBaseMomentProcess","Current time: " + result_msg)
+
+    def _GetCorrespondingBaseMoment(self):
+        return KratosCFD.DragUtilities().CalculateBodyFittedBaseMoment(self.model_part,self.reference_point)
+        # return KratosCFD.DragUtilities().CalculateBodyFittedDrag(self.model_part)
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/python_scripts/compute_body_fitted_base_moment_process.py:32]: Remove commented code.

File: applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp [hunk 7]
```diff
@@ -118,6 +118,90 @@ namespace Kratos
...
+    array_1d<double, 3> DragUtilities::CalculateEmbeddedBaseMoment(ModelPart& rModelPart, array_1d<double, 3> rReferencePoint) {
+        // Initialize total drag force
+        array_1d<double, 3> base_moment = ZeroVector(3);
+        double& moment_x = base_moment[0];
+        double& moment_y = base_moment[1];
+        double& moment_z = base_moment[2];
+
+        // Iterate the model part elements to compute the moments
+        array_1d<double, 3> elem_drag; 
+        array_1d<double, 3> elem_drag_center;
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp:165]:
```suggestion
        array_1d<double, 3> elem_base_moment; 
        array_1d<double, 3> elem_base_moment_center;
```

File: applications/FluidDynamicsApplication/custom_utilities/base_moment_utilities.h [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+    * of the given modelpart and the base moment.
+    * @param rModelPart reference to the model part in where the drag is to be computed
+    * @return An array containing the drag force value and the base moments.
+    */
+    array_1d<double, 3> CalculateBodyFittedBaseMoment(ModelPart &rModelPart, array_1d<double, 3> rReferencePoint);
+
+    /**
+    * Computes the integral of the Cauchy stress term normal projection in the given modelpart elements.
+    * @param rModelPart reference to the model part in where the drag is to be computed
+    * @return An array containing the drag force value and base moments.
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/custom_utilities/base_moment_utilities.h]:
```suggestion
    * @param rModelPart reference to the model part in where the base moment is to be computed
    * @return An array containing the base moment value and base moments.
```

File: applications/FluidDynamicsApplication/custom_utilities/base_moment_utilities.h [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+
+
+    ///@}
+    ///@name Operations
+    ///@{
+
+    /**
+    * Computes the integral of the pressure stress term normal projection over the conditions
+    * of the given modelpart and the base moment.
+    * @param rModelPart reference to the model part in where the drag is to be computed
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/custom_utilities/base_moment_utilities.h]:
```suggestion
    * @param rModelPart reference to the model part in where the base moment is to be computed
```

File: applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp [hunk 10]
```diff
@@ -118,6 +118,90 @@ namespace Kratos
         return drag_force_center;
     }
 
+
+    array_1d<double, 3> DragUtilities::CalculateBodyFittedBaseMoment(ModelPart& rModelPart, array_1d<double, 3> rReferencePoint) {
```
Comment (riccardotosi) [inline applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp:122]: Isn't this function the same of `base_moment_utilities.cpp`?

**Review by riccardotosi — COMMENTED** [review]
A test is also missing.

**Review by philbucher — COMMENTED** [review]
overall looks good but I have some comment, will post in the next days

File: applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp (hunk 10 above)
Comment (akodakkal) [reply applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp:122]: @rubenzorrilla Is it ok to include this functionality in drag_utilities.cpp or a separate base_moment_utilities is required ?

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp [hunk 11]
```diff
@@ -234,6 +234,82 @@ namespace Kratos {
...
+            ref_point(0) = 0.0;
+            ref_point(1) = 0.0;
+            ref_point(2) = 0.0;
+
+            // Call the body fitted drag utility
+            DragUtilities drag_utilities;
+            array_1d<double, 3> base_moment = drag_utilities.CalculateBodyFittedBaseMoment(model_part.GetSubModelPart("DragModelPart"), ref_point);
+
+            // Check computed values
+            KRATOS_WATCH_VECTOR_WITH_PRECISION(base_moment, 16);
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp:271]:
```suggestion
```
I am quite sure this won't compile without cotire

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp [hunk 12]
```diff
@@ -234,6 +234,82 @@ namespace Kratos {
...
+            ref_point(0) = 0.0;
+            ref_point(1) = 0.0;
+            ref_point(2) = 0.0;
+
+            // Call the embedded drag utility
+            DragUtilities drag_utilities;
+            array_1d<double, 3> base_moment = drag_utilities.CalculateEmbeddedBaseMoment(model_part, ref_point);
+
+            // Check computed values
+            KRATOS_WATCH_VECTOR_WITH_PRECISION(base_moment, 16);
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp:306]:
```suggestion
```

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp [hunk 13]
```diff
@@ -234,6 +234,82 @@ namespace Kratos {
...
+
+            // Call the body fitted drag utility
+            DragUtilities drag_utilities;
+            array_1d<double, 3> base_moment = drag_utilities.CalculateBodyFittedBaseMoment(model_part.GetSubModelPart("DragModelPart"), ref_point);
+
+            // Check computed values
+            KRATOS_WATCH_VECTOR_WITH_PRECISION(base_moment, 16);
+            KRATOS_CHECK_NEAR(base_moment[0], 0.0, 1e-6);
+            KRATOS_CHECK_NEAR(base_moment[1], 0.0, 1e-6);
+            KRATOS_CHECK_NEAR(base_moment[2], 40.0, 1e-6);
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp:274]: why such low precision?

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp [hunk 14]
```diff
@@ -234,6 +234,82 @@ namespace Kratos {
...
+
+            // Call the embedded drag utility
+            DragUtilities drag_utilities;
+            array_1d<double, 3> base_moment = drag_utilities.CalculateEmbeddedBaseMoment(model_part, ref_point);
+
+            // Check computed values
+            KRATOS_WATCH_VECTOR_WITH_PRECISION(base_moment, 16);
+            KRATOS_CHECK_NEAR(base_moment[0], 0.0, 1e-2);
+            KRATOS_CHECK_NEAR(base_moment[1], 0.0, 1e-4);
+            KRATOS_CHECK_NEAR(base_moment[2], 0.238125, 1e-6);
```
Comment (philbucher) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_drag_utils.cpp:309]: same

File: applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp (hunk 10 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_utilities/drag_utilities.cpp:122]: As I discussed with @riccardotosi in the office, in here I see two options:
1. If it has to be a separated `base_moment_utilities` file, I'd definitively move it to the `ExaquteSandboxApplication` so we get rid of maintaining it once the project finishes.
2. Considering that this utility may eventually be used for other purposes (i.e. calculating the pitching moment) we could think about including it in the `FluidDynamicsApplication`.

Taking into account that the `FluidDynamicsApplication` is supposed to be a core application, for option 2 we would need to:
- Use the `parallel_utils`
- Follow the style guide standards (removing commented code, indentation, ...)
- Change the name of the utility and methods to remove the `Base...` (this is misleading as the moment is computed with respect to a user-defined point).
- Enhance testing. If I'm not wrong, right now we check an in-plane drag force with (0,0,0) as reference point. This might eventually lead to false positiveness of the test.
- We should definitively not include 3 extra Python scripts to the application. For this we need to discuss about how do we want to output the drag and the moment. Most probably we want to merge everything in one process.

Taking everything into account, my personal view is to include it as it is right now in the application that was specifically created for the project so we don't interfere with the project. In here I'd like to note that the `FluidDynamicsApplication` is a hard to maintain application which is going through a (slow) clean up process. At some point we may think about rescuing it from the project app.

**Comment by loumalouomega** [comment]
@akodakkal can we delete the branch?
